File: examples/dekita/.claude/hooks/git-status-check.py
#!/usr/bin/env python3
"""セッション開始時にGitの状態を確認し、未コミット変更を警告する。

Why:
    mainブランチに未コミット変更がある状態でセッションを開始すると、
    意図しない変更の混入や競合が発生するリスクがある。

What:
    - Gitワーキングツリーの状態を確認
    - mainブランチに未コミット変更がある場合に警告
    - featureブランチの未コミット変更は情報として表示

Remarks:
    - ブロックはしない（警告のみ）
    - stop_hook_active時はスキップ
"""

import json
import subprocess
import sys

from lib.constants import TIMEOUT_LIGHT
from lib.execution import log_hook_execution
from lib.session import parse_hook_input


def get_git_status() -> tuple[bool, str, str]:
    """
    Check git status and branch.
    Returns (is_clean, branch_name, status_output).
    """
    try:
        # Get current branch
        branch_result = subprocess.run(
            ["git", "branch", "--show-current"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_LIGHT,
        )
        branch = branch_result.stdout.strip()

        # Get git status
        status_result = subprocess.run(
            ["git", "status", "--porcelain"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_LIGHT,
        )
        status_output = status_result.stdout.strip()
        is_clean = len(status_output) == 0

        return is_clean, branch, status_output
    except subprocess.TimeoutExpired:
        return True, "", "timeout"
    except Exception as e:
        return True, "", str(e)


def main():
    """
    Entry point for the git status check hook.

    Reads a JSON object from stdin with the following format:
        {
            "transcript_path": "<path to transcript file>",
            "stop_hook_active": <bool, optional>
        }

    Outputs a JSON object to stdout with the following format:
        {
            "decision": "approve",
            "reason": "<explanation>",
            "systemMessage": "<message to display>"
        }
    """
    result = None
    try:
        input_json = parse_hook_input()

        # If stop_hook_active is set, approve immediately
        if input_json.get("stop_hook_active"):
            result = {
                "ok": True,
                "decision": "approve",
                "reason": "stop_hook_active is set; skipping git status check.",
            }
        else:
            is_clean, branch, status_output = get_git_status()

            if is_clean:
                result = {
                    "ok": True,
                    "decision": "approve",
                    "reason": f"Git working tree is clean on branch '{branch}'",
                    "systemMessage": f"✅ git-status-check: クリーン ({branch})",
                }
            elif branch == "main":
                # Warning: uncommitted changes on main branch
                result = {
                    "ok": True,
                    "decision": "approve",  # Don't block, just warn
                    "reason": f"Uncommitted changes on main branch:\n{status_output}",
                    "systemMessage": f"⚠️ git-status-check: mainブランチに未コミット変更があります\n{status_output}",
                }
            else:
                # Changes on feature branch - OK
                result = {
                    "ok": True,
                    "decision": "approve",
                    "reason": f"Uncommitted changes on branch '{branch}' (not main, OK)",
                    "systemMessage": f"ℹ️ git-status-check: 未コミット変更あり ({branch})",
                }

    except Exception as e:
        # On error, approve to avoid blocking
        print(f"[git-status-check] Hook error: {e}", file=sys.stderr)
        result = {"ok": True, "decision": "approve", "reason": f"Hook error: {e}"}

    log_hook_execution("git-status-check", result.get("decision", "approve"), result.get("reason"))
    print(json.dumps(result))
    sys.exit(0)


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/git_status_check.ts
#!/usr/bin/env bun
/**
 * セッション開始時にGitの状態を確認し、未コミット変更を警告する。
 *
 * Why:
 *   mainブランチに未コミット変更がある状態でセッションを開始すると、
 *   意図しない変更の混入や競合が発生するリスクがある。
 *
 * What:
 *   - Gitワーキングツリーの状態を確認
 *   - mainブランチに未コミット変更がある場合に警告
 *   - featureブランチの未コミット変更は情報として表示
 *
 * Remarks:
 *   - ブロックはしない（警告のみ）
 *   - stop_hook_active時はスキップ
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#2874: TypeScript移行
 *   - silenvx/dekita#3094: async化（spawnSync → asyncSpawn）
 */

import { TIMEOUT_LIGHT } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { parseHookInput } from "../lib/session";
import { asyncSpawnAll } from "../lib/spawn";

const HOOK_NAME = "git-status-check";

export interface GitStatus {
  isClean: boolean;
  branch: string;
  statusOutput: string;
}

/**
 * Check git status and branch (async, parallel execution).
 * @internal Exported for testing
 */
export async function getGitStatus(): Promise<GitStatus> {
  try {
    // Execute both commands in parallel
    const [branchResult, statusResult] = await asyncSpawnAll([
      {
        command: "git",
        args: ["branch", "--show-current"],
        options: { timeout: TIMEOUT_LIGHT * 1000 },
      },
      {
        command: "git",
        args: ["status", "--porcelain"],
        options: { timeout: TIMEOUT_LIGHT * 1000 },
      },
    ]);

    // gitコマンド失敗時はエラーとして扱う（isClean: falseでエラーを隠蔽しない）
    if (!branchResult.success) {
      return {
        isClean: false,
        branch: "",
        statusOutput: `git branch --show-current failed: ${branchResult.stderr || "Unknown error"}`,
      };
    }
    if (!statusResult.success) {
      return {
        isClean: false,
        branch: "",
        statusOutput: `git status --porcelain failed: ${statusResult.stderr || "Unknown error"}`,
      };
    }

    const branch = branchResult.stdout?.trim() || "";
    const statusOutput = statusResult.stdout?.trim() || "";
    const isClean = statusOutput.length === 0;

    return { isClean, branch, statusOutput };
  } catch (e) {
    // エラー時もisClean: falseとし、エラーを隠蔽しない
    return { isClean: false, branch: "", statusOutput: `error: ${formatError(e)}` };
  }
}

interface HookResult {
  ok: boolean;
  decision?: string;
  reason?: string;
  systemMessage?: string;
}

async function main(): Promise<void> {
  let result: HookResult;
  let sessionId: string | undefined;

  try {
    const inputJson = await parseHookInput();
    sessionId = inputJson.session_id;

    // If stop_hook_active is set, approve immediately
    if (inputJson.stop_hook_active) {
      result = {
        ok: true,

        reason: "stop_hook_active is set; skipping git status check.",
      };
    } else {
      const { isClean, branch, statusOutput } = await getGitStatus();

      if (isClean) {
        result = {
          ok: true,

          reason: `Git working tree is clean on branch '${branch}'`,
          systemMessage: `✅ git-status-check: クリーン (${branch})`,
        };
      } else if (branch === "main") {
        // Warning: uncommitted changes on main branch
        result = {
          ok: true,
          // Don't block, just warn
          reason: `Uncommitted changes on main branch:\n${statusOutput}`,
          systemMessage: `⚠️ git-status-check: mainブランチに未コミット変更があります\n${statusOutput}`,
        };
      } else {
        // Changes on feature branch - OK
        result = {
          ok: true,

          reason: `Uncommitted changes on branch '${branch}' (not main, OK)`,
          systemMessage: `ℹ️ git-status-check: 未コミット変更あり (${branch})`,
        };
      }
    }
  } catch (e) {
    // On error, approve to avoid blocking
    console.error(`[${HOOK_NAME}] Hook error: ${formatError(e)}`);
    result = { ok: true, reason: `Hook error: ${formatError(e)}` };
  }

  await logHookExecution(HOOK_NAME, result.decision ?? "approve", result.reason, undefined, {
    sessionId,
  });
  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
